File: lib/keywords.js
/**
 * Resolver: maps natural-language descriptions to traits via the trigger map
 * defined in content/resolvers/triggers.json.
 *
 * Two tiers (per the locked tier rule in triggers.json):
 *   high_confidence: phrase implies traits unconditionally → auto-apply.
 *   low_confidence:  phrase implies traits only sometimes → return as a
 *                    suggestion with a confirmation_prompt; the caller decides
 *                    whether to add the trait based on user input.
 *
 * Matching: case-insensitive substring match against the input text. Multi-word
 * phrases match contiguously. The same source file is read by the CLI here AND
 * embedded into IDE rules files by `aigis init` (see lib/init.js), so agents
 * and the CLI use the exact same map.
 */

const fs = require('fs');
const path = require('path');

const TRIGGERS_PATH = path.join(__dirname, '..', 'content', 'resolvers', 'triggers.json');

let _cache = null;
function loadTriggers() {
  if (!_cache) _cache = JSON.parse(fs.readFileSync(TRIGGERS_PATH, 'utf8'));
  return _cache;
}

/**
 * Scan `text` against the high-confidence trigger map. Returns:
 *   {
 *     high_confidence_matches: [
 *       { phrase: 'chatbot', traits: ['uses-llm', 'accepts-user-input'] },
 *       ...
 *     ],
 *     low_confidence_matches: [
 *       { id: '1', phrase: 'personalization', traits: ['processes-pii'],
 *         confirmation_prompt: '...', notes: '...' },
 *       ...
 *     ],
 *     traits_auto_applied: ['uses-llm', 'accepts-user-input', ...]    // unique union
 *   }
 *
 * The caller is responsible for handling the low_confidence_matches array
 * (interactive prompt, --confirm/--reject flags, default-reject in non-TTY).
 */
function detectTraitsFromText(text) {
  const triggers = loadTriggers();
  const textLower = text.toLowerCase();

  const highMatches = [];
  const autoTraits = new Set();
  for (const [phrase, entry] of Object.entries(triggers.high_confidence)) {
    if (textLower.includes(phrase.toLowerCase())) {
      highMatches.push({ phrase, traits: entry.traits.slice() });
      for (const t of entry.traits) autoTraits.add(t);
    }
  }

  const lowMatches = [];
  let nextId = 1;
  for (const [phrase, entry] of Object.entries(triggers.low_confidence)) {
    if (textLower.includes(phrase.toLowerCase())) {
      lowMatches.push({
        id: String(nextId++),
        phrase,
        traits: entry.traits.slice(),
        confirmation_prompt: entry.confirmation_prompt,
        notes: entry.notes,
      });
    }
  }

  return {
    high_confidence_matches: highMatches,
    low_confidence_matches: lowMatches,
    traits_auto_applied: [...autoTraits],
  };
}

/**
 * Resolve a list of low-confidence suggestions against user decisions.
 * decisions = { '1': 'yes' | 'no' | 'unsure', ... }  (id-keyed)
 * Returns the union of high-confidence + confirmed low-confidence traits.
 */
function applyConfirmations(detection, decisions = {}) {
  const final = new Set(detection.traits_auto_applied);
  const userDecisions = [];
  for (const sug of detection.low_confidence_matches) {
    const decision = (decisions[sug.id] || 'no').toLowerCase();
    userDecisions.push({ ...sug, user_decision: decision });
    if (decision === 'yes' || decision === 'y') {
      for (const t of sug.traits) final.add(t);
    }
  }
  return {
    final_traits: [...final].sort(),
    low_confidence_decisions: userDecisions,
  };
}

module.exports = { detectTraitsFromText, applyConfirmations, loadTriggers, TRIGGERS_PATH };

File: cli-python/aigis_cli/keywords.py
"""Mirror of lib/keywords.js — the resolver.

Loads content/resolvers/triggers.json and provides:
  - detect_traits_from_text(text): returns high/low matches + auto-applied traits
  - apply_confirmations(detection, decisions): merges low-confidence yes answers

The two-tier match logic is identical to the JS implementation.
"""

from __future__ import annotations

import json
from pathlib import Path

CONTENT_DIR = Path(__file__).resolve().parent / "content"
TRIGGERS_PATH = CONTENT_DIR / "resolvers" / "triggers.json"

_cache: dict | None = None


def load_triggers() -> dict:
    global _cache
    if _cache is None:
        _cache = json.loads(TRIGGERS_PATH.read_text(encoding="utf-8"))
    return _cache


def detect_traits_from_text(text: str) -> dict:
    triggers = load_triggers()
    text_lower = text.lower()

    high_matches = []
    auto_traits: set[str] = set()
    for phrase, entry in triggers["high_confidence"].items():
        if phrase.lower() in text_lower:
            high_matches.append({"phrase": phrase, "traits": list(entry["traits"])})
            for t in entry["traits"]:
                auto_traits.add(t)

    low_matches = []
    next_id = 1
    for phrase, entry in triggers["low_confidence"].items():
        if phrase.lower() in text_lower:
            low_matches.append({
                "id": str(next_id),
                "phrase": phrase,
                "traits": list(entry["traits"]),
                "confirmation_prompt": entry["confirmation_prompt"],
                "notes": entry["notes"],
            })
            next_id += 1

    return {
        "high_confidence_matches": high_matches,
        "low_confidence_matches": low_matches,
        "traits_auto_applied": list(auto_traits),
    }


def apply_confirmations(detection: dict, decisions: dict | None = None) -> dict:
    decisions = decisions or {}
    final = set(detection["traits_auto_applied"])
    user_decisions = []
    for sug in detection["low_confidence_matches"]:
        decision = (decisions.get(sug["id"], "no") or "no").lower()
        user_decisions.append({**sug, "user_decision": decision})
        if decision in ("yes", "y"):
            for t in sug["traits"]:
                final.add(t)
    return {
        "final_traits": sorted(final),
        "low_confidence_decisions": user_decisions,
    }
